src/Aggregate/SnapshottingAggregateRoot.php

```diff
@@ -0,0 +1,30 @@
+<?php
+declare(strict_types=1);
+
+namespace MyOnlineStore\EventSourcing\Aggregate;
+
+use MyOnlineStore\EventSourcing\Event\Stream;
+use MyOnlineStore\EventSourcing\Service\Assert;
+
+abstract class SnapshottingAggregateRoot extends AggregateRoot
+{
+    public static function reconstituteFromSnapshotAndHistory(
+        Snapshot $snapshot,
+        Stream $eventStream
+    ): SnapshottingAggregateRoot {
+        $instance = \unserialize(\base64_decode($snapshot->getState()));
```

Comment:
You could add the `allowed_classes` option here to assert the result always is an instance of the implementation.

Reply:
That's not really possible, because we'd have to account for all classes that are used within an aggregate (All VOs, collections, sub-entities, etc).